### PR TITLE
Implement xexxar’s aim assessment

### PR DESCRIPTION
**Pull request by smoogipoo**
Prereqs:
- [x] https://github.com/ppy/osu/pull/4099

Implements the changes which were discussed over the last 1-2 months in https://github.com/ppy/osu-performance/issues/68 and #osu-performance.

Includes speed balancing changes.

### CODE REVIEW

**Comment by Xexxar** [comment]
Originally when balancing the buff to aim, I chose `single_spacing_threshold` rather arbitrarily. Considering that my original intention was to nerf speed values exceeding `stream_spacing_threshold`, it probably makes sense that aim should receive buffs for wide angles exceeding `stream_spacing_threshold`. I believe this change will probably balance wide angles better. The `.5x` buff might be a little large with this value, however. 

With that being said, once such a change is done, I suggest implementing this change. While waiting for the length rework is valid, I've become disenfranchised and frustrated with the limitations that the SR system has by using 400 MS chunks. I believe I have a method that could be used to replace the 400 MS chunks system, but this would take much longer to implement and would require substantial work in rebalancing and testing. In addition, the change would have nothing to with the actual difficulty curves themselves, so there would be no change to the altercations in this PR.

For this reason, since this change will be felt across a much smaller group of players and the effects are overall seen a nearly pure positive, pushing this change to live seems to be the best move right now. Endlessly waiting when changes are ready to be implemented and enjoyed seems silly, especially during the holiday season. There may be cases where abuse can occur, but with the state of the current system, not implementing this change seems even more irresponsible than doing so. 

I hope you all understand where I am coming from.

**Comment by PowerChaos125** [comment]
> 
> 
> Originally when balancing the buff to aim, I chose `single_spacing_threshold` rather arbitrarily. Considering that my original intention was to nerf speed values exceeding `stream_spacing_threshold`, it probably makes sense that aim should receive buffs for wide angles exceeding `stream_spacing_threshold`. I believe this change will probably balance wide angles better. The `.5x` buff might be a little large with this value, however.
> 

I think this is not an elegant solution for what you want to do. By this change, we will have a part of jump distance from 110 to 125 that is counted in both speed and aim.
I suggest 2 solutions
- Make speed scale to 110 rather than 125 and make aim scale start from 110 as you suggest
- Keep the current distance threshold, and increase aim angle bonus multiplier.

**Comment by VINXIS** [comment]
Xexxar and i spent 10 hrs today this better work

Aim https://puu.sh/Clmj2/e28d0d8c15.cs
Speed https://puu.sh/Clmje/8a2b18c88a.cs

didnt make it nicer sry....

**Comment by VINXIS** [comment]
In Aim.cs

```cs
result = 2 * Math.Pow(Math.Max(0, angleBonus), 0.99) / Math.Max(Previous[0].StrainTime, timing_threshold);
```
the Max function for strain should instead be:
```cs
result = 2 * Math.Pow(Math.Max(0, angleBonus), 0.99) / Math.Max(timing_threshold, Math.Max(current.StrainTime, Previous[0].StrainTime));```

my bad

**Comment by VINXIS** [comment]
Also as Xexxar stated an HD change needs to be implemented to rebalance the HD mod with these changes, his suggestion was the following changes to **OsuPerformanceCalculator.cs**, which was tested in #osu-performance a few hours ago today with the speed and angle changed implemented:

aimValue *= 1.0f + 0.04f * (12.0f - Attributes.ApproachRate) | Line 122
speedValue *= 1.0f + 0.04f * (12.0f - Attributes.ApproachRate) | Line 154
accValue *= 1.08f | Line 187

**Comment by VINXIS** [comment]
Note: https://github.com/ppy/osu-performance/issues/73#issuecomment-449599322

File: osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
Comment (VINXIS) [inline osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs]: This is now broken due to the removal of `computeSliderCursorPosition(lastSlider);`

**Review by VINXIS — CHANGES_REQUESTED** [review]

**Comment by VINXIS** [comment]
In Angle Bonus please multiply `Previous[0].TravelDistance` by 8 Line 36 Aim.cs

**Comment by VINXIS** [comment]
Not sure if Xexxar approves but I made a small change to aim.cs

Here are comparisons to whats on xexxar-all and the change i made:
https://hastebin.com/xikakalode.apache nm
https://hastebin.com/japifawika.apache hr
https://hastebin.com/xabuxufudi.apache dt

after
https://hastebin.com/evawodebel.apache nm
https://hastebin.com/piviyonera.apache hr
https://hastebin.com/zonocaheta.apache dt

```cs
(current.JumpDistance - scale, 0)
```

turns into

```cs
(current.JumpDistance + 8 * current.TravelDistance - scale, 0)
```

in Aim.cs's anglebonus calc

Waiting for Xexxar's Approval

**Comment by VINXIS** [comment]
Xexxar approved

**Comment by VINXIS** [comment]
I'm sorry smoogi one more change....


in Speed.cs
```cs
return (1 + (speedBonus-1) * .75) * angleBonus * (0.95 + speedBonus * Math.Pow(distance / SINGLE_SPACING_THRESHOLD, 4)) / current.StrainTime;
```

**Comment by smoogipoo** [comment]
I'm calling this one complete now. Thanks to people in #osu-performance, who've gone through the ranked + unranked beatmaps comparing lazer (smoogipoo/diffcalc-fixes) vs xexxar (this pr).

Ranked: https://docs.google.com/spreadsheets/d/1NHeXCXz16rT-y1ymgvoHoLZITe3GH6VQJj2n35dwA_M/edit?ouid=107198598249851885548&usp=sheets_home&ths=true

Unranked: https://docs.google.com/spreadsheets/d/1CkXLOYsPi67aV04JmIwSsvvYTbJHHGrlutEyDjhn_Wg/edit#gid=0

It seems like most cases are able to be justified as either being silly maps (tag4/2b/etc), or having wide angle jumps (eg. 1735958).

File: osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs
Comment (peppy) [reply osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs]: Has this been resolved?

**Review by peppy — APPROVED** [review]

**Pull request merged**
